shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/EraIndepGenerators.hs

```diff
@@ -442,21 +444,35 @@ instance (Era era, Mock (Crypto era)) => Arbitrary (DPState era) where
   shrink = genericShrink
 
 instance
-  (ShelleyBased era, Mock (Crypto era), Arbitrary (Core.Value era)) =>
+  (ShelleyBased era, Mock (Crypto era), Arbitrary (Core.Value era),
+  Arbitrary (State (Core.UpdateSTS era)) ) =>
   Arbitrary (UTxOState era)
   where
   arbitrary = genericArbitraryU
-  shrink = genericShrink
+  shrink = recursivelyShrink
+    -- The 'genericShrink' function returns first the immediate subterms of a
+    -- value (in case it is a recursive data-type), and then shrinks the value
+    -- itself. Since 'UTxOState' is not a recursive data-type, there are no
+    -- subterms, and we can use `recursivelyShrink` directly. This is
+    -- particularly important when abstracting away the different fields of the
+    -- ledger state, since the generic subterms instances will overlap due to
+    -- GHC not having enough context to infer if 'a' and 'b' are the same types
+    -- (since in this case this will depend on the definition of 'era').
+    --
+    -- > instance OVERLAPPING_ GSubtermsIncl (K1 i a) a where
+    -- > instance OVERLAPPING_ GSubtermsIncl (K1 i a) b where
```

Comment:
I want to write a heart picture here, but I don't know how to do the little pictures. Thanks so much for including this great explanation!

Reply:
Credit here goes to Nick Frisby, who explained this to me.